src/scanner.cc

```diff
@@ -1,346 +1,228 @@
-    // consume '%' * multistring_level
-    // consume '{' * 1
-    for (unsigned int i = 0; i < (multistring_level + 1); i++) {
+    if (lookahead(lexer) == 'm') {
+      m = true;
       advance(lexer);
     }
-    lexer->mark_end(lexer);
-    return true;
+
+    expected_percent_count.pop_back();
+
+    // An END is fully scanned when we started with an '"' (precondition of
+    // this function), consumed all %-signs and ended with an m.
+    return (m && count == 0);
   }
-  DEBUG_PRINT("scan_multistring_richterm_start: slow path\n");
-  // %{ comes after this
-  // consume all the % required in this multistring
-  for (unsigned int i = 0; i < multistring_level; i++) {
-    if (lexer->lookahead != '%') {
-      return false;
-    }
+
+  // Precondition of this function is that the lookahead is '"'
+  bool scan_str_start(TSLexer *lexer) {
+    lexer->result_symbol = STR_START;
+
+    // Interpolation in strings are preceded by a single % sign.
+    expected_percent_count.push_back(1);
+
     advance(lexer);
+
+    return true;
   }
-  if (lexer->lookahead != '{') {
-    return false;
+
+  // Precondition of this function is that the lookahead is '"'
+  bool scan_str_end(TSLexer *lexer) {
+    lexer->result_symbol = STR_END;
+
+    advance(lexer);
+
+    expected_percent_count.pop_back();
+
+    return true;
   }
-  advance(lexer);
-  lexer->mark_end(lexer);
-  return true;
-}
 
-static bool scan_multistring_end(TSLexer *lexer) {
-  DEBUG_PRINT("scan_multistring_end\n");
-  lexer->result_symbol = MULTISTRING_END;
-  // "% was parsed before this
-  // expected: "%m "%%m "%%%m "%%%%m ...
-  // multistring_level = number of %
-  for (unsigned int i = 0; i < (multistring_level - 1); i++) {
-    if (lexer->lookahead == '%') {
+  bool scan_interpolation_start(TSLexer *lexer) {
+    lexer->result_symbol = INTERPOLATION_START;
+
+    bool brace = false;
+    uint8_t count = expected_percent_count.back();
+
+    while (lookahead(lexer) == '%') {
+      count--;
       advance(lexer);
-      DEBUG_PRINT("scan_multistring_end: found %% %i of %i\n", (i+1), (multistring_level - 1));
     }
-    else {
-      DEBUG_PRINT("scan_multistring_end: found %c but expected %% -> return false\n", lexer->lookahead);
-      return false;
+
+    if (lookahead(lexer) == '{') {
+      brace = true;
+      advance(lexer);
     }
+
+    return brace && (count == 0);
   }
-  if (lexer->lookahead == 'm') {
+
+  // Precondition of this function is that the lookahead is '}'
+  bool scan_interpolation_end(TSLexer *lexer) {
+    lexer->result_symbol = INTERPOLATION_END;
+
     advance(lexer);
-    DEBUG_PRINT("scan_multistring_end: found m -> return true\n");
+
     return true;
   }
-  DEBUG_PRINT("scan_multistring_end: not found m -> return false\n");
-  return false;
-}
 
+  bool scan(TSLexer *lexer, const bool *valid_symbols) {
+    while (iswspace(lookahead(lexer))) {
+      skip(lexer);
+    }
+
+    if (lookahead(lexer) == 'm' && valid_symbols[MULTSTR_START]) {
```

Comment:
use `switch` here? so compiler can generate a jump table

Reply:
I don't think a jump table can be generated either way because of the `valid_symbols` constraint. Also a switch wouldn't be very nice with each case having a nested if.

Reply:
benchmark to the rescue ; ) but i guess the difference is near zero